receiver/sapmreceiver/trace_receiver.go

```diff
@@ -0,0 +1,239 @@
+	TraceEndpointV2 = "/v2/trace"
+
+	traceSource string = "sapm"
+
+	contentTypeHeader = "Content-Type"
+	xprotobuf         = "application/x-protobuf"
+
+	acceptEncodingHeader  = "Accept-Encoding"
+	contentEncodingHeader = "Content-Encoding"
+	gzipEncoding          = "gzip"
+)
+
+var gzipWriterPool = &sync.Pool{
+	New: func() interface{} {
+		return gzip.NewWriter(ioutil.Discard)
+	},
+}
+
+// sapmReceiver receives spans in the Splunk SAPM format over HTTP
+type sapmReceiver struct {
+	// mu protects the fields of this type
+	mu        sync.Mutex
+	startOnce sync.Once
+	stopOnce  sync.Once
+	logger    *zap.Logger
+
+	config *Config
+	server *http.Server
+
+	nextConsumer consumer.TraceConsumer
+}
+
+// handleRequest parses an http request containing sapm and passes the trace data to the next consumer
+func (sr *sapmReceiver) handleRequest(ctx context.Context, req *http.Request) error {
+	sapm, err := sapmrequestparser.ParseTraceV2Request(req)
+	// errors processing the request should return http.StatusBadRequest
+	if err != nil {
+		return err
+	}
+
+	// pass the SAPM and error to the receiver function
+	// process the sapm
+	if sapm == nil {
+		return errors.New("incoming sapm was nil")
+	}
+
+	// process sapm batches
+	for _, batch := range sapm.Batches {
+		// convert the jager batches to OCProto
+		td, err := jaegertranslator.ProtoBatchToOCProto(*batch)
+		if err != nil {
+			return err
+		}
+		td.SourceFormat = "sapm"
+
+		// pass the trace data to the next consumer
+		err = sr.nextConsumer.ConsumeTraceData(ctx, td)
+		if err != nil {
+			return fmt.Errorf("error passing trace data to next consumer: %v", err.Error())
+		}
+
+		// We MUST unconditionally record metrics from this reception.
+		observability.RecordMetricsForTraceReceiver(ctx, len(batch.Spans), len(batch.Spans)-len(td.Spans))
+	}
+
+	return nil
+}
+
+// HTTPHandlerFunction returns an http.HandlerFunc that handles SAPM requests
+func (sr *sapmReceiver) HTTPHandlerFunc(rw http.ResponseWriter, req *http.Request) {
+	// create context with the receiver name from the request context
+	ctx := observability.ContextWithReceiverName(req.Context(), "sapm")
+
+	// trace this request
+	_, span := trace.StartSpan(ctx, traceSource)
+	defer span.End()
+
+	// handle the request payload
+	err := sr.handleRequest(ctx, req)
+	if err != nil {
+		// TODO account for this error (throttled logging or metrics)
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// respBytes are bytes to write to the http.Response
+	var respBytes []byte
+
+	// build the response message
+	respBytes, err = proto.Marshal(&splunksapm.PostSpansResponse{})
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	rw.Header().Set(contentTypeHeader, xprotobuf)
+
+	// write the response if client does not accept gzip encoding
+	if req.Header.Get(acceptEncodingHeader) != gzipEncoding {
+		// write the response bytes
+		rw.Write(respBytes)
+		return
+	}
+
+	// gzip the response
```

Comment:
I think the thinking is that in the future this might return something.  I'll go ahead and make a default response in the receiver's constructor and send that for now.

Reply:
I think we should do it if acceptEndocingHeader is present. At least it documents what the behavior should be. As an optimization, until we actually need a real response, we can have the client never ask for gzipped response. Either that or the encoding code below can just check if the response to be sent is empty and skip encoding.

Reply:
Ok cool, then I'm just going to have a default response in []byte.  And we'll still pass that through the gzip code here.  If we ever decide to put stuff in the response we can revise this function to actually build the response byte array before going through this gzip path.